✨  Add KS multi-cluster monitoring with Prometheus & Thanos

dumb0002 opened the pull request:
<!--
Thanks for creating a pull request!

If this is your first time, please make sure to review CONTRIBUTING.MD.

Please copy the appropriate `:text:` or icon to the beginning of your PR title:

:sparkles: ✨ feature
:bug: 🐛 bug fix
:book: 📖 docs
:memo: 📝 proposal
:warning: ⚠️ breaking change
:seedling: 🌱 other/misc
:question: ❓ requires manual review/categorization

-->
## Summary

This PR adds KS multi-cluster monitoring with Thanos, more precisely :

-   Add scripts to install Thanos with Minio storage for long term persistence of data in the core cluster
-   Configure Prometheus deployed in any cluster (i.e., core or wec) to remote write to Thanos
-   Update Grafana dashboard to visualize monitoring data for the KS components running in either core or WEC clusters (single pane of glass for monitoring kS components.) 
-  Add scripts to install KS monitoring tools in the WEC
-  Update the diagram illustrating KS monitoring architecture. 

## Related issue(s)

Fixes #

MikeSpreitzer commented on monitoring/README.md:
```diff
@@ -4,38 +4,67 @@
 
```
I think it would be good to use the regular terminology for the two status add-on components. See https://docs.kubestellar.io/release-0.24.0/direct/architecture/ . One is called the status add-on controller and the other is the status add-on agent.

dumb0002 replied:
Fixed - updated the architecture image to use that terminology.

MikeSpreitzer commented on monitoring/images/ks-monitoring-setup.png:
I see that in the new architecture, the backing store for all the data --- except the core's pyroscope --- is the core's Minio. Why the exception? Would it be better to store the core's pyroscope data in Minio too?

dumb0002 replied:
Yes! you're correct - I missed to show that in the architecture diagram - core's pyroscope should store the data in Minio too - I just updated the new architecture and I will push all the pyroscope configuration in a new PR.

MikeSpreitzer commented on monitoring/images/ks-monitoring-setup.png:
Why is the WEC's kube-apiserver not subject to pyroscope profiling?

dumb0002 replied:
Do we need to collect profile data for the kube-apiservers? It's important to collect profile information for our KS controllers as it will help us to debug and improve the implementation of our controller. However, it is not clear to me the need to collect profile data for the kube-apiservers.

MikeSpreitzer commented on monitoring/README.md:
```diff
@@ -4,38 +4,67 @@
 
```
I would be comforted if the the picture said specifically "MinIO _Server_", because we have used "MinIO" as a shorthand for MinIO _client_ --- and even as a shorthand for the S3 protocol --- in the past.

dumb0002 replied:
I have seen that when the word "MinIO" is used it implies the server and a clear distinction is used when referring to the MinIO client. For example, see the reference for Minio server and client docker images below:   

- Minio Server:  https://hub.docker.com/r/minio/minio
- Minio Client: 
     i) https://hub.docker.com/r/bitnami/minio-client/
    ii) https://hub.docker.com/r/minio/mc/

MikeSpreitzer commented on monitoring/README.md:
````diff
@@ -4,38 +4,67 @@
 
+a) Deploy the monitoring tools in the hosting cluster (e.g., `kind-kubeflex`):
+
+```bash
+$ cd monitoring
+$ ./install-ks-monitoring.sh --cluster-context kind-kubeflex --set core
+```
````
I see a "copy" icon for this displayed text, but doing so is useless because of the included command prompts (the dollar signs).

<img width="887" alt="image" src="https://github.com/user-attachments/assets/a767f46a-bfec-430f-a744-1d83101a891f">

dumb0002 replied:
Github adds the copy item in the README by default, is it?

MikeSpreitzer replied:
Let me try an experiment here.

```shell
echo my command
```

```console
$ echo my command
my command
```

MikeSpreitzer replied:
I was hoping that using `console` rather than `shell` as the [info-string](https://github.github.com/gfm/#info-string) (based on the list found from https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#syntax-highlighting) would produce different rendering, but experiment shows "nope".

MikeSpreitzer replied:
FYI, you _could_ use the style used in the rest of our documentation, where every fenced block is _either_ command input (with prompt omitted) or command output --- but not a mixture of both. That way, readers can simply use the copy button on the command input blocks.

dumb0002 replied:
Done! Fixed

MikeSpreitzer reviewed (commented):
I tested, and it failed as follows. First I used `test/e2e/run-test.sh` to create a working system. Then I started following the instructions in this PR.

```shell
mspreitz@mjs13 kubestellar % cd monitoring 
mspreitz@mjs13 monitoring % 
mspreitz@mjs13 monitoring % ./install-ks-monitoring.sh --cluster-context kind-kubeflex --set core 
+ set -e
+ ctx=kind-kubeflex
+ ns=ks-monitoring
+ opt=core
+ '[' 4 '!=' 0 ']'
+ case "$1" in
+ ((  4 > 1  ))
+ ctx=kind-kubeflex
+ shift
+ shift
+ '[' 2 '!=' 0 ']'
+ case "$1" in
+ ((  2 > 1  ))
+ opt=core
+ shift
+ shift
+ '[' 0 '!=' 0 ']'
+ case "$opt" in
+++ dirname -- ./install-ks-monitoring.sh
++ cd -- .
++ pwd
+ SCRIPT_DIR=/Users/mspreitz/go/src/github.com/kubestellar/kubestellar/monitoring
+ source /Users/mspreitz/go/src/github.com/kubestellar/kubestellar/monitoring/setup-shell.sh
++ export -f wait-for-object
++ export -f wait-for-cmd
+ /Users/mspreitz/go/src/github.com/kubestellar/kubestellar/monitoring/cleanup.sh --cluster-context kind-kubeflex --ns ks-monitoring
+ set -e
+ ns=ks-monitoring
+ ctx=kind-kubeflex
+ '[' 4 '!=' 0 ']'
+ case "$1" in
+ ((  4 > 1  ))
+ ctx=kind-kubeflex
+ shift
+ shift
+ '[' 2 '!=' 0 ']'
+ case "$1" in
+ ((  2 > 1  ))
+ ns=ks-monitoring
+ shift
+ shift
+ '[' 0 '!=' 0 ']'
+ helm --kube-context kind-kubeflex delete -n ks-monitoring grafana --ignore-not-found
release "grafana" uninstalled
+ helm --kube-context kind-kubeflex delete -n ks-monitoring pyroscope --ignore-not-found
release "pyroscope" uninstalled
+ helm --kube-context kind-kubeflex delete -n ks-monitoring prometheus --ignore-not-found
release "prometheus" uninstalled
+ helm --kube-context kind-kubeflex delete -n ks-monitoring minio --ignore-not-found
release "minio" uninstalled
+ helm --kube-context kind-kubeflex delete -n ks-monitoring thanos --ignore-not-found
release "thanos" uninstalled
+ kubectl --context kind-kubeflex -n ks-monitoring delete pvc --all
No resources found
+ kubectl --context kind-kubeflex delete ns ks-monitoring --ignore-not-found
+ : --------------------------------------------------------------
+ : Create KS monitoring namespace
+ kubectl config use-context kind-kubeflex
Switched to context "kind-kubeflex".
+ kubectl create namespace ks-monitoring
namespace/ks-monitoring created
+ '[' core == core ']'
+ : --------------------------------------------------------------
+ : Install Thanos with MinIO storage
+ : Install MinIO MinIO object storage for Thanos to store metrics
+ helm repo add minio https://charts.min.io/
"minio" has been added to your repositories
+ helm install minio -n ks-monitoring -f /Users/mspreitz/go/src/github.com/kubestellar/kubestellar/monitoring/prometheus/minio-custom-values.yaml minio/minio --version 5.2.0
Error: INSTALLATION FAILED: failed post-install: 1 error occurred:
	* timed out waiting for the condition


mspreitz@mjs13 monitoring % 
```

MikeSpreitzer commented:
The minio pods are in Pending STATUS.

```shell
mspreitz@mjs13 monitoring % kubectl --context kind-kubeflex describe pod -n ks-monitoring minio-0
Name:             minio-0
Namespace:        ks-monitoring
Priority:         0
Service Account:  minio-sa
Node:             <none>
Labels:           app=minio
                  apps.kubernetes.io/pod-index=0
                  controller-revision-hash=minio-768f575778
                  release=minio
                  statefulset.kubernetes.io/pod-name=minio-0
Annotations:      checksum/config: 54dcefc396ddf736bce30789a84739042217f69914f1bd4a0c9ae22e7283ddc1
                  checksum/secrets: 3606e402ecb75a6c11a8b1a81f5bb6b8f8ec1315ee49a48b21583b5a12edfd11
Status:           Pending
IP:               
IPs:              <none>
Controlled By:    StatefulSet/minio
Containers:
  minio:
    Image:       quay.io/minio/minio:RELEASE.2024-04-18T19-09-19Z
    Ports:       9000/TCP, 9001/TCP
    Host Ports:  0/TCP, 0/TCP
    Command:
      /bin/sh
      -ce
      /usr/bin/docker-entrypoint.sh minio server http://minio-{0...1}.minio-svc.ks-monitoring.svc.cluster.local/export -S /etc/minio/certs/ --address :9000 --console-address :9001
    Requests:
      memory:  16Gi
    Environment:
      MINIO_ROOT_USER:             <set to the key 'rootUser' in secret 'minio'>      Optional: false
      MINIO_ROOT_PASSWORD:         <set to the key 'rootPassword' in secret 'minio'>  Optional: false
      MINIO_PROMETHEUS_AUTH_TYPE:  public
    Mounts:
      /export from export (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-hbtr9 (ro)
Conditions:
  Type           Status
  PodScheduled   False 
Volumes:
  export:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  export-minio-0
    ReadOnly:   false
  minio-user:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  minio
    Optional:    false
  kube-api-access-hbtr9:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason            Age                  From               Message
  ----     ------            ----                 ----               -------
  Warning  FailedScheduling  3m55s (x3 over 11m)  default-scheduler  0/1 nodes are available: 1 Insufficient memory. preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod.
```

MikeSpreitzer commented:
`kubectl --context kind-kubeflex get Node kubeflex-control-plane -o yaml` shows that the one and only Node in my `kind`-based KubeFlex hosting cluster has 11 GB of memory.

dumb0002 commented on monitoring/README.md:
````diff
@@ -4,38 +4,67 @@
+b) Deploy the monitoring tools in a WEC cluster (e.g., `cluster1`):
+
 ```bash
 $ cd monitoring
````
Yes! it would work. We don't need to change directory indeed - it was a typo! Fixed it now

dumb0002 commented:
> `kubectl --context kind-kubeflex get Node kubeflex-control-plane -o yaml` shows that the one and only Node in my `kind`-based KubeFlex hosting cluster has 11 GB of memory.

@MikeSpreitzer, I figured the issue. I was creating very large pv for minio/thanos/prometheus - I am using a test VM with enough resource. Hence, I decreased significantly the resources requirements and you should now be able to run in your laptop.

MikeSpreitzer commented:
I am testing commit 8126af48e6dc745bc3fd5cd1c3b9fd5494e2e61e --- the latest rev of this PR. Still the minio pods are requesting 16GiB of memory.

dumb0002 commented:
> I am testing commit [8126af4](https://github.com/kubestellar/kubestellar/commit/8126af48e6dc745bc3fd5cd1c3b9fd5494e2e61e) --- the latest rev of this PR. Still the minio pods are requesting 16GiB of memory. I mean memory, not a PV.
> 
> [pr2458-try2.log](https://github.com/user-attachments/files/17066203/pr2458-try2.log)

I just pushed a commit with changes to decreased the memory from 16G to 2G. Also, I run few tests and everything seems to be working as expected with that amount of memory.  The minio memory can now be configured in this file: monitoring/prometheus/minio-custom-values.yaml

MikeSpreitzer commented:
I tested the latest version of this PR, commit e29e26a019491347b18426eaf07778f90ff9e729. I first used `test/e2e/run-test.sh` to create a working system. Then:

```console
mspreitz@mjs13 kubestellar % cd monitoring 
mspreitz@mjs13 monitoring % 
mspreitz@mjs13 monitoring % ./install-ks-monitoring.sh --cluster-context kind-kubeflex --set core
+ set -e
+ ctx=kind-kubeflex
...
+ : Install Thanos
+ helm -n ks-monitoring upgrade --install thanos bitnami/thanos --values /Users/mspreitz/go/src/github.com/kubestellar/kubestellar/monitoring/prometheus/thanos-custom-values.yaml --version 15.7.25
Release "thanos" does not exist. Installing it now.
Error: repo bitnami not found
```

dumb0002 commented:
> I tested the latest version of this PR, commit [e29e26a](https://github.com/kubestellar/kubestellar/commit/e29e26a019491347b18426eaf07778f90ff9e729). I first used `test/e2e/run-test.sh` to create a working system. Then:
> 
> ```
> mspreitz@mjs13 kubestellar % cd monitoring 
> mspreitz@mjs13 monitoring % 
> mspreitz@mjs13 monitoring % ./install-ks-monitoring.sh --cluster-context kind-kubeflex --set core
> + set -e
> + ctx=kind-kubeflex
> ...
> + : Install Thanos
> + helm -n ks-monitoring upgrade --install thanos bitnami/thanos --values /Users/mspreitz/go/src/github.com/kubestellar/kubestellar/monitoring/prometheus/thanos-custom-values.yaml --version 15.7.25
> Release "thanos" does not exist. Installing it now.
> Error: repo bitnami not found
> ```

@MikeSpreitzer, oh! Sorry I was missing to add the thanos helm chart repo. Just pushed a commit to fix it. It should work as expected now. Thanks!

MikeSpreitzer commented on monitoring/README.md:
````diff
@@ -4,43 +4,83 @@
 
 ```bash
-$ kubectl -n ks-monitoring get servicemonitor -l "app.kubernetes.io/part-of!=kube-prometheus-stack" | grep wds1*
+kubectl -n ks-monitoring get servicemonitor -l "app.kubernetes.io/part-of!=kube-prometheus-stack" | grep wds1*
````
This fails due to the lack of quoting around `wds1*`.

dumb0002 replied:
Ok, I see! Fixed but it works for me even without any quotes:

```
$ kubectl -n ks-monitoring get servicemonitor -l "app.kubernetes.io/part-of!=kube-prometheus-stack" | grep wds1*
wds1-apiserver                 3d19h
wds1-ks-controller-manager     3d19h
wds1-ks-transport-controller   3d19h
```

```
$ kubectl -n ks-monitoring get servicemonitor -l "app.kubernetes.io/part-of!=kube-prometheus-stack" | grep its1*
its1-apiserver                 3d19h
its1-status-addon-controller   3d19h
```

MikeSpreitzer commented on monitoring/README.md:
````diff
@@ -49,26 +89,51 @@ wds1-ks-transport-controller   64m
+./configure-monitoring-wec.sh --wec-cluster-context cluster1
+```
+
+Optionally, check the prometheus service monitor object:
````
I think your intent is that this "Optionally..." is inside the body of the above loop over "every WEC", but textually that is not clear; this "Optionally..." appears _after_ the loop and at the same level of indentation as the loop.

dumb0002 replied:
I added some new text to make it clear that the "optionally" statement is intent for every WEC. Also added the `--context` flag to avoid any ambiguity.

MikeSpreitzer commented on monitoring/README.md:
````diff
@@ -49,26 +89,51 @@ wds1-ks-transport-controller   64m
 
 ```bash
-$ kubectl port-forward -n ks-monitoring svc/grafana 3000:80
+kubectl port-forward -n ks-monitoring svc/grafana 3000:80
````
```console
mspreitz@mjs13 monitoring % kubectl port-forward -n ks-monitoring svc/grafana 3000:80
Error from server (NotFound): services "grafana" not found
```

dumb0002 replied:
Fixed! Added the `--context` flag as above!

MikeSpreitzer commented on monitoring/README.md:
````diff
@@ -49,26 +89,51 @@ wds1-ks-transport-controller   64m
+kubectl port-forward -n ks-monitoring svc/grafana 3000:80
 ```
 
 b) Check Prometheus and Pyroscope data source configuration: in a browser, go to the Grafana server at http://localhost:3000 and on the left-hand side, go to Home > Connections > Data sources.
````
I am testing. I see the data sources. When I import the dashboards, all the panels say "no data". Looking more directly at the Thanos data source, it also says "no data".

<img width="917" alt="image" src="https://github.com/user-attachments/assets/9c5ee019-4472-4f3a-9b7b-960f66c45232">

MikeSpreitzer commented on monitoring/README.md:
```diff
@@ -92,47 +157,108 @@ c) Upload the dashboard JSON files from your local `./grafana` sub-directory int
 a) API Server Monitoring: input into the following text boxes.
 - `NS`: name of the KubeStellar space namespace, e.g., *wds1-system*, *wds2-system*, *its1-system*, etc.
 - `APIService`: name of the service for a KubeStellar space APIserver, e.g., *wds1*, *wds2*, *vcluster*, etc.
+- `Instance`: hostname and port number for the target KubeStellar space APIserver, e.g., *10.244.0.18:9444*
```
At this point the user should be told how to find the right values for "Instance".

MikeSpreitzer replied:
This turns up a lot:

```console
mspreitz@mjs13 ~ % kubectl --context kind-kubeflex get Endpoints -A | egrep "NAME| wds| vcluster |kubernetes"
NAMESPACE         NAME                                                              ENDPOINTS                                               AGE
default           kubernetes                                                        172.19.0.2:6443                                         97m
its1-system       vcluster                                                          10.244.0.14:8443,10.244.0.14:8443                       94m
wds1-system       wds1                                                              10.244.0.17:9444                                        93m
```

dumb0002 replied:
Fixed! Using the simpler way provided above. Thanks!

MikeSpreitzer commented on monitoring/README.md:
```diff
@@ -92,47 +157,108 @@ c) Upload the dashboard JSON files from your local `./grafana` sub-directory int
+- `Instance`: hostname and port number for the target KubeStellar space APIserver, e.g., *10.244.0.18:9444*
 
     <img src="images/apiserver-monitoring.png" width="60%" height="80%" title="apiserver-monitoring"> 
 
```
FYI: the panels using PromQL's `histogram_quantile` function have spurious precision (because that function does), and you could help the reader by listing the histogram bucket boundaries in the panel's description.

dumb0002 replied:
Could you open an issue for this feature improvement? I would like to address this into a new PR. Many thanks!

MikeSpreitzer replied:
Done. See #2466 .

dumb0002 replied:
Thank you!

MikeSpreitzer commented on monitoring/README.md:
````diff
@@ -92,47 +157,108 @@ c) Upload the dashboard JSON files from your local `./grafana` sub-directory int
+   - You can also use the following combination of `NS/APIService` values to monitor the APIserver of the hosting cluster or a WEC cluster: `NS: default` and `APIService: apiserver`
+   - Use the following commands to obtain the Ip address of the APIServer in the WDS (e.g., `wds1`) and ITS (e.g., `its1-system`) spaces:
+     ```bash
+        kubectl -n wds1-system get pod -l app=kube-apiserver -o jsonpath="{.items[*].status.podIP}"
````
Wrong kubeconfig context.

dumb0002 replied:
Fixed!

MikeSpreitzer commented on monitoring/README.md:
```diff
@@ -92,47 +157,108 @@ c) Upload the dashboard JSON files from your local `./grafana` sub-directory int
+
+Likewise, to monitor the KubeStellar controller in a ITS space and WEC (e.g., its1): `NS: its1-system` and `KSController: status-addon-controller` & `NS: open-cluster-management-agent-addon` and `KSController: status-agent-controller`.
+
+Note: 
```
This turns up a lot:

```console
mspreitz@mjs13 ~ % kubectl --context kind-kubeflex get Endpoints -A | egrep "NAME|system .*controller"
NAMESPACE         NAME                                                              ENDPOINTS                                               AGE
its1-system       status-addon-controller                                           10.244.0.51:9280                                        63m
kubeflex-system   kubeflex-controller-manager-metrics-service                       10.244.0.11:8443                                        117m
wds1-system       ks-transport-controller                                           10.244.0.49:8090                                        64m
wds1-system       kubestellar-controller-manager-metrics-service                    10.244.0.48:8443                                        113m
```

dumb0002 replied:
Fixed! I am using the namespace flag to restrict to the relevant namespaces only (e.g., `wds1-system` & `its1-system`).

MikeSpreitzer replied:
My point was about using Endpoints rather than a more tedious approach.

dumb0002 replied:
Yes, I already updated the instructions to use the Endpoints approach. Thanks!

MikeSpreitzer commented on monitoring/README.md:
```diff
@@ -92,47 +157,108 @@ c) Upload the dashboard JSON files from your local `./grafana` sub-directory int
-- `KSController`: select the KubeStellar controller name. Available options from the drop-down menu: *kubestellar-controller-manager-metrics-service*, *ks-transport-controller* and *status-addon-controller*. 
+- `KSController`: select the KubeStellar controller name. Available options from the drop-down menu: *kubestellar-controller-manager-metrics-service*, *ks-transport-controller*, *status-addon-controller* and *status-agent-controller*. 
+
+- `Instance`: hostname and port number for the target KubeStellar space controller, e.g., *10.244.0.71:8090*
```
Do we really need to make the user supply this value? In other words, why include "instance" in the filter? Isn't there always exactly one right "instance" given the other two user inputs?

dumb0002 replied:
No, that's not the always the case! For example: the apiserver in the hosting cluster & in the WEC cluster have exactly the same namespace (`default`) and job name (`apiserver`) and the only distinction is the `instance` value

MikeSpreitzer commented on monitoring/README.md:
````diff
@@ -92,47 +157,113 @@ c) Upload the dashboard JSON files from your local `./grafana` sub-directory int
+
+     For example: 
+     ```bash
+     docker inspect -f '{{range.NetworkSettings.Networks}}{{.IPAddress}}{{end}}' cluster1-control-plane
````
Endpoints make this job easier too.

```console
mspreitz@mjs13 monitoring % kubectl --context cluster1 get Endpoints -n default 
NAME         ENDPOINTS         AGE
kubernetes   172.19.0.3:6443   25m
```

I see a different port number here (6443) than below (9444). In my testing just now, 6443 gets me data and 9444 does not.

Yes, it is the same IP address as from the `docker inspect`:

```console
mspreitz@mjs13 monitoring % docker inspect -f '{{range.NetworkSettings.Networks}}{{.IPAddress}}{{end}}' cluster1-control-plane
172.19.0.3
```

dumb0002 replied:
Agreed! I updated to use endpoints instead of docker specific commands. Yes, port `# 6443`  is the correct one - the list of correct ports are also listed in the README instructions. 

`9444` is probably the port for your docker container itself.

MikeSpreitzer commented on monitoring/README.md:
````diff
@@ -92,47 +157,113 @@ c) Upload the dashboard JSON files from your local `./grafana` sub-directory int
+     wds1-system       wds1                                                              10.244.0.21:9444                                         3d20h
+     ```
+     
+     If you're using kind clusters then use the following command to obtain the IP address of the host container for your cluster: `$ docker inspect -f '{{range.NetworkSettings.Networks}}{{.IPAddress}}{{end}}' <docker-container-name>`
````
The reader may wonder why she wants "the IP address of the host container for your cluster".  Nothing in this document says anything about using that value, right?

dumb0002 replied:
I updated the instructions to use endpoints instead as discussed in the previous comments above.

MikeSpreitzer commented on monitoring/README.md:
```diff
@@ -92,47 +157,113 @@ c) Upload the dashboard JSON files from your local `./grafana` sub-directory int
     <img src="images/apiserver-apf-monitoring.png" width="60%" height="80%" title="apiserver-apf-monitoring.png"> 
 
-Note: use the same combinations of `NS` and `APIService` parameters described in `step 5-a` to monitor the APIserver in the WDS space, ITS space or the hosting cluster. 
+**Note**: use the same combinations of `NS` and `APIService` parameters described in `step 5-a` to monitor the APIserver in the WDS space, ITS space, the hosting or WEC clusters. 
```
`instance` too, right?

dumb0002 replied:
Yes, but the instance value is not constant - it's environment dependent. These instructions just explain the combination of constant values that the user should or must use.

MikeSpreitzer replied:
But the user needs to know what value to use for the "instance" variable. Something should be said about that. If I understand correctly, the same combination of `NS`, `APIService`, and `Instance` that works for a given target in 5-a also works for the same target in 5-b. Couldn't the instructions say something that simple, without getting tangled up in the fact that the discussion in 5-a is structured as a table plus some other comments?

dumb0002 replied:
I created an issue to keep track of this here: https://github.com/kubestellar/kubestellar/issues/2470 - I will address it in a new PR. Thanks!

MikeSpreitzer commented on monitoring/README.md:
```diff
@@ -92,47 +157,113 @@ c) Upload the dashboard JSON files from your local `./grafana` sub-directory int
+- `NS`: name of the KubeStellar space namespace, e.g., *wds1-system*, *wds2-system*, etc.
 
-- `KSController`: select the KubeStellar controller name. Available options from the drop-down menu: *kubestellar-controller-manager-metrics-service*, *ks-transport-controller* and *status-addon-controller*. 
+- `KSController`: select the KubeStellar controller name. Available options from the drop-down menu: *kubestellar-controller-manager-metrics-service*, *ks-transport-controller*, *status-addon-controller* and *status-agent-controller*. 
```
"status-agent-controller" is not a thing. The "OCM Status add-on agent" _is_ a thing.

MikeSpreitzer replied:
My point is not about spaces vs dashes. My point is that the architecture document does not talk about a "status agent controller" (with any punctuation); I think the intended thing is what the architecture document calls the "OCM Status Add-on Agent". I am OK with dropping "OCM" from the name in the menu. But putting "controller" in there is not what we want, I think.

dumb0002 replied:
I will address this in a new PR. I created an issue to keep track of this here: https://github.com/kubestellar/kubestellar/issues/2470

MikeSpreitzer commented:
I am testing the latest rev of this PR. I have gotten to the controllers dashboard, am trying to configure for WECs. But I am getting no data from my WECs.

<img width="845" alt="image" src="https://github.com/user-attachments/assets/af132b64-4f90-4529-929d-7493f04c1b63">

MikeSpreitzer commented:
I am testing the latest version of this PR. I have gotten to the last case: the controllers dashboard, the case of the status add-on agent in the WEC. Thanos has no data for the alleged namespace for that:

<img width="847" alt="image" src="https://github.com/user-attachments/assets/851bce4d-6c63-4a32-9c9d-aca0a3c96c55">

MikeSpreitzer commented:
Looking directly at the Prometheus server in cluster1, I see that it does not hold `workqueue_depth` timeseries from anything but the apiserver.

<img width="841" alt="image" src="https://github.com/user-attachments/assets/53dc33e8-8e3d-4d9c-8a7e-2cbb95401117">

MikeSpreitzer commented:
The Prometheus server in cluster1 is not even trying to get metrics from anything but the apiserver.

<img width="847" alt="image" src="https://github.com/user-attachments/assets/5224b8e3-2e2f-4984-8dd1-f99c58896cc3">

MikeSpreitzer commented:
That Service has zero Endpoints:

```console
mspreitz@mjs13 ~ % kubectl --context cluster1 get Endpoints -n open-cluster-management-agent-addon status-agent-controller -o yaml
apiVersion: v1
kind: Endpoints
metadata:
  annotations:
    endpoints.kubernetes.io/last-change-trigger-time: "2024-09-24T05:34:19Z"
  creationTimestamp: "2024-09-24T05:34:19Z"
  labels:
    app: status-agent
  name: status-agent-controller
  namespace: open-cluster-management-agent-addon
  resourceVersion: "3088"
  uid: 27769a96-04b0-48cc-8f29-759397cafa7a
```

MikeSpreitzer commented:
The agent Pod has no ports declared, I looked at the output of `kubectl get pods -n open-cluster-management-agent-addon -l app=status-agent -o yaml` (long, omitted here for brevity).

But:

```console
mspreitz@mjs13 ~ % kubectl --context cluster1 get services -n open-cluster-management-agent-addon -l app=status-agent -o yaml
apiVersion: v1
items:
- apiVersion: v1
  kind: Service
  metadata:
    ...
    name: status-agent-controller
    namespace: open-cluster-management-agent-addon
    ...
  spec:
    ...
    ports:
    - name: metrics
      port: 8080
      protocol: TCP
      targetPort: metrics
...
```

dumb0002 commented:
@MikeSpreitzer (cc: @pdettori), here are the instructions on how to deploy the status-addon-controller and status-addon-agent. It will address most of your comments above:

PR: https://github.com/kubestellar/ocm-status-addon/pull/83

**Steps on how to test the status-addon-controllers using the repo helm chart:**

   1 - Remove the status-addon-controller and status-addon-agent from your current environmet:

```
helm --kube-context its1 -n open-cluster-management delete status-addon
k --context its1 -n cluster1 delete manifestwork addon-addon-status-deploy-0
```

   2 - Clone my fork:

``` bash
 git clone -b metrics-scrapping https://github.com/dumb0002/ocm-status-addon.git
```
  3  - Build the chart:

```
cd ocm-status-addon/
make build
make chart
```

 4  - Build the images locally and push to kind image registry:

```
cd ocm-status-addon/
make ko-local-build
make kind-load-image CLUSTERS="kubeflex cluster1 cluster2"
```

5 - Deploy the status-on-controllers using the local chart:

```
make deploy DEFAULT_IMBS_CONTEXT="its1"
```

dumb0002 commented:
@MikeSpreitzer, this PR is merged https://github.com/kubestellar/ocm-status-addon/pull/83 and a new release is also available. Hence, you can now follow these two simple steps to re-deploy the status-addon-agent and status-addon-controller in your env:

1.  Remove the status-addon-controller and status-addon-agent from your current environment:

```
helm --kube-context its1 -n open-cluster-management delete status-addon
k --context its1 -n cluster1 delete manifestwork addon-addon-status-deploy-0
```

2. Deploy the status-on-controllers: 

```
helm --kube-context its1 upgrade --install ocm-status-addon -n open-cluster-management oci://ghcr.io/kubestellar/ocm-status-addon-chart --version v0.2.0-rc12
```

ezrasilvera commented:
@dumb0002 I see many open comments and then later on answers claiming some are fixed and so on.  Some of the new instructions (status addon etc..) are spread across comment answers, this is hard to follow. Can you please point me to the updated instructions to test this so I can try it? 
Let's try to test and if successful I suggest to close all comments related to issues in running and then you can focus on the other comments (functionality, doc phrasing, etc..)
I also recommend to move improvement suggestion to follow-up PRs otherwise it will be difficult to converge (especially on the test validation)

dumb0002 commented:
Thanks for your comments @ezrasilvera! You can find the updated instructions to test it here: https://github.com/dumb0002/kubestellar/tree/thanos-configuration/monitoring

MikeSpreitzer commented:
As far as I can tell, nobody has followed the instructions in https://docs.kubestellar.io/release-0.24.0/direct/release/#reacting-to-a-new-ocm-status-addon-release . We should do that first, then no special hackery would be needed here.

MikeSpreitzer commented on monitoring/README.md:
````diff
@@ -3,44 +3,101 @@
+
+```bash
+helm --kube-context its1 -n open-cluster-management delete status-addon
+k --context its1 -n cluster1 delete manifestwork addon-addon-status-deploy-0
````
Please write out `kubectl` --- aliasing it to `k` is not something that we have told users that they have to do.

dumb0002 replied:
Fixed!

MikeSpreitzer commented on monitoring/README.md:
```diff
@@ -3,44 +3,101 @@
 
 
+### Requirement
+To follow the instructions here, you need to have release `v0.2.0-rc12` or higher for the status-addon controller and agent installed in your enviroment. If you're using an older release, please follow these instructions to update your environment: 
```
Rather than do a special hack here, we should first follow the instructions in https://docs.kubestellar.io/release-0.24.0/direct/release/#reacting-to-a-new-ocm-status-addon-release . After that, no special hackery would be needed here.

dumb0002 replied:
This is just a temporary hack until we have a new release for KS.

MikeSpreitzer replied:
But someone using `main` no longer needs this hack, right? (#2468 has merged.)

dumb0002 replied:
To address that - I updated the README file with the following sentence: 
```
To follow the instructions here you need to use KubeStellar main branch or any future KubeStellar release higher than 0.24.0
```
I also created this issue to keep track of improvements needed in the instructions: https://github.com/kubestellar/kubestellar/issues/2470

MikeSpreitzer commented on monitoring/README.md:
````diff
@@ -3,44 +3,101 @@
+helm --kube-context its1 upgrade --install ocm-status-addon -n open-cluster-management oci://ghcr.io/kubestellar/ocm-status-addon-chart --version v0.2.0-rc12
+```
+
 ### Description
````
The reader should be told what versions of KubeStellar this is intended to work with. Release 0.24.0? Branch `main`?

dumb0002 replied:
I added the following information in the README: 
```
To follow the instructions here you need to use KubeStellar main branch for any future KubeStellar release higher than `0.24.0`
```
This is just a point in time instructions. Ideally, these instructions will be intended for specific release of KS - to conduct performance analysis for a specific release of KS. However, right now it works also with `main` - but this branch changes constantly.

MikeSpreitzer commented on monitoring/README.md:
```diff
@@ -3,44 +3,101 @@
+minio-0                                                1/1     Running   0          3m58s
+minio-1                                                1/1     Running   0          3m58s
+prometheus-kube-prometheus-operator-79d5cf844f-nbzvb   1/1     Running   0          2m57s
+prometheus-prometheus-kube-prometheus-prometheus-0     3/3     Running   0          2m54s
```
FYI, I see 2/2 instead of 3/3 here when I test this.

dumb0002 replied:
This is correct! Probably information from an old configuration. Just updated it with the output of the latest configuration that matches your output.

MikeSpreitzer commented on monitoring/README.md:
```diff
@@ -92,47 +174,124 @@ c) Upload the dashboard JSON files from your local `./grafana` sub-directory int
+| Host-cluster  |     6443      |
+| WDS space     |     6443      |
+| ITS space     |     8443      |
+| WEC-cluster   |     9444      |
```
I test the system produced by `test/e2e/run-test.sh`. In that system, in the WECs, the apiserver port number (the subject of this table) is 6443.

dumb0002 replied:
That was a typo. The correct port is `6443` indeed - updated the README file accordingly

MikeSpreitzer commented on monitoring/README.md:
````diff
@@ -92,47 +174,124 @@ c) Upload the dashboard JSON files from your local `./grafana` sub-directory int
+     kubernetes   172.18.0.3:6443   4d14h
+     ```
+
+Here are the default port number for the APIservers: 
````
I am not sure "default" is on point here. The reader needs to know how to determine the correct actual port number for their actual running system. The command exhibited just above does this for WEC apiservers. Other commands further above cover other rows in this table. Is this table needed?

dumb0002 replied:
This table is just a helper table for the user. The information showed there are the typical information for an environment created using kind clusters.

MikeSpreitzer commented on monitoring/README.md:
````diff
@@ -92,47 +174,124 @@ c) Upload the dashboard JSON files from your local `./grafana` sub-directory int
+  kubectl --context kind-kubeflex -n wds1-system get Endpoints | egrep ".*controller"
+  ```
+  ```bash
+  kubectl --context kind-kubeflex -n its1-system get Endpoints | egrep ".*controller"
````
I am testing the latest rev of this PR (commit b249c02a8e45eadcfbf088d288f91cb48cb114fb). I get "no data" in two of the six panels in the status-addon-controller case.

<img width="880" alt="image" src="https://github.com/user-attachments/assets/1c2c5b7b-efc9-4017-939a-1a25c250398b">

dumb0002 replied:
This is correct - this is relative to the missing metrics issues from our controllers that I reported in the past. Also, see this issue: https://github.com/kubestellar/kubestellar/issues/2158

MikeSpreitzer replied:
I think it would be good if this document tells the reader not to expect data in those panels.

MikeSpreitzer commented on monitoring/README.md:
````diff
@@ -92,47 +174,124 @@ c) Upload the dashboard JSON files from your local `./grafana` sub-directory int
+  kubectl --context kind-kubeflex -n its1-system get Endpoints | egrep ".*controller"
+  ```
+  ```bash
+  kubectl --context cluster1 -n open-cluster-management-agent-addon get Endpoints
````
I get data in 3 and "no data" in 3 of the 6 panels when testing the cluster1 case.

<img width="892" alt="image" src="https://github.com/user-attachments/assets/948f51c3-5fc4-46d9-b0b7-f72a44868373">

dumb0002 replied:
This is correct - this is related to the missing metrics issues for some of our controllers that I reported in the past. Also, see this issue: https://github.com/kubestellar/kubestellar/issues/2158

MikeSpreitzer replied:
I think that it would be best if the reader were told what to expect.

MikeSpreitzer commented on monitoring/README.md:
````diff
@@ -92,47 +174,124 @@ c) Upload the dashboard JSON files from your local `./grafana` sub-directory int
+  kubectl --context kind-kubeflex -n its1-system get Endpoints | egrep ".*controller"
+  ```
+  ```bash
+  kubectl --context cluster1 -n open-cluster-management-agent-addon get Endpoints
````
What stands my two WECs apart?

```console
mspreitz@mjs13 kubestellar % kubectl --context cluster1 -n open-cluster-management-agent-addon get Endpoints
NAME                      ENDPOINTS         AGE
status-agent-controller   10.244.0.9:8080   31m

mspreitz@mjs13 kubestellar % kubectl --context cluster2 -n open-cluster-management-agent-addon get Endpoints
NAME                      ENDPOINTS         AGE
status-agent-controller   10.244.0.9:8080   33m
```

dumb0002 replied:
I will address this in a new PR - I created an issue to keep track of this here: https://github.com/kubestellar/kubestellar/issues/2470

MikeSpreitzer commented on monitoring/README.md:
```diff
@@ -92,47 +174,124 @@ c) Upload the dashboard JSON files from your local `./grafana` sub-directory int
 
     <img src="images/ks-controllers-pyroscope.png" width="60%" height="80%" title="ks-controllers-pyroscope">
 
+c) Query profiles for the Kubestellar controller in a WEC cluster: connect to the UI of the Pyroscope application running in the target WEC, since there is not a single pane of glass to visualize Pyroscope data across all clusters in our monitoring architecture:
```
What does "the Kubestellar controller in a WEC" mean? That's not something in our architecture document.

Does a WEC cluster have a PIN number?

dumb0002 replied:
I meant the `status-agent-controller`.  

What do you mean by a PIN number?

MikeSpreitzer replied:
It was a joking way to point out a redundancy. "WEC" is "Workload Execution Cluster", so "WEC cluster" is redundant.

MikeSpreitzer reviewed (commented):
I have finished another round of review.

dumb0002 commented on monitoring/README.md:
````diff
@@ -3,44 +3,101 @@
+```bash
+NAME                                                   READY   STATUS    RESTARTS   AGE
+prometheus-kube-prometheus-operator-79d5cf844f-nbzvb   1/1     Running   0          2m57s
+prometheus-prometheus-kube-prometheus-prometheus-0     3/3     Running   0          2m54s
````
This is correct! Probably information from an old configuration. Just updated it with the output of the latest configuration that matches your output.

dumb0002 commented:
> I have finished another round of review.

@MikeSpreitzer, I addressed all your comments. I also successfully tested these instructions with some minor modifications in a OCP env. I will create a new PR to include the instructions on how to deploy into a OCP cluster (cc: @ezrasilvera )

MikeSpreitzer commented on monitoring/README.md:
```diff
@@ -92,47 +175,124 @@ c) Upload the dashboard JSON files from your local `./grafana` sub-directory int
-    ii) Click `Explore` in the Pyroscope datasource selection
+   i) Click `Connections -> Data Sources` in the primary menu
+  ii) Click `Explore` in the Pyroscope datasource selection
 
```
FYI, I get a lot of failures when I try the Pyroscope stuff. They seem to be transient; I see no container restarts or Pod replacements. But here is what I see in my shell running the main port-forward. The errors come only from the Pyroscope part.

```console
mspreitz@mjs13 monitoring % kubectl --context kind-kubeflex port-forward -n ks-monitoring svc/grafana 3000:80
Forwarding from 127.0.0.1:3000 -> 3000
Forwarding from [::1]:3000 -> 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
E0929 21:22:00.412418   51571 portforward.go:381] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:3000->127.0.0.1:52873: write tcp4 127.0.0.1:3000->127.0.0.1:52873: write: connection reset by peer
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
E0929 21:22:36.976687   51571 portforward.go:381] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:3000->127.0.0.1:52890: write tcp4 127.0.0.1:3000->127.0.0.1:52890: write: broken pipe
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
E0929 21:23:06.621547   51571 portforward.go:370] error creating forwarding stream for port 3000 -> 3000: Timeout occurred
E0929 21:23:06.972897   51571 portforward.go:347] error creating error stream for port 3000 -> 3000: Timeout occurred
E0929 21:23:07.088161   51571 portforward.go:370] error creating forwarding stream for port 3000 -> 3000: Timeout occurred
E0929 21:23:07.090158   51571 portforward.go:347] error creating error stream for port 3000 -> 3000: Timeout occurred
Handling connection for 3000
E0929 21:23:28.066825   51571 portforward.go:370] error creating forwarding stream for port 3000 -> 3000: Timeout occurred
Handling connection for 3000
E0929 21:23:49.321673   51571 portforward.go:347] error creating error stream for port 3000 -> 3000: Timeout occurred
Handling connection for 3000
E0929 21:25:00.695454   51571 portforward.go:370] error creating forwarding stream for port 3000 -> 3000: Timeout occurred
E0929 21:25:42.538100   51571 portforward.go:381] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:3000->127.0.0.1:52896: write tcp4 127.0.0.1:3000->127.0.0.1:52896: write: broken pipe
Handling connection for 3000
Handling connection for 3000
E0929 21:26:19.237958   51571 portforward.go:347] error creating error stream for port 3000 -> 3000: Timeout occurred
```

MikeSpreitzer commented on monitoring/README.md:
````diff
@@ -92,47 +175,124 @@ c) Upload the dashboard JSON files from your local `./grafana` sub-directory int
+
+  ```bash
+  kubectl --context cluster1 -n ks-monitoring port-forward svc/pyroscope 9090:4040
+  ```
````
FYI, I think that it would be helpful at this point to explicitly tell the user to direct a browser to http://localhost:9090/ (readers might not remember whether the 9090 or the 4040 is the port on the local side).

MikeSpreitzer reviewed (commented):
If you want to relegate the remaining doc issues to #2470 then this PR is good enough, it seems to be as functional as expected.
/lgtm

dumb0002 commented:
/approve
